prov/rxm: Add PID to CM messages

shefty opened the pull request:
There are 2 situations where we can receive a connection request
from a peer where we already have a connection that's active.

In one case, the connection may be an older connection to the
same node, but to a client that's no longer running.  In this case,
the old connection needs to be closed and the new connection
accepted.  This is the behavior of the current code.  The code
was structured this way to handle a situation where a client
crashed and was restarted, with the new client choosing the same
network address.  Unless the old connection discarded and the new
request accepted, the re-started client would not be able to
communicate with a running server.

In the second case, the connection may have just been created,
with the request being for a simultaneous connection.  This
occurs frequently during MPI startup when all to all communication
is needed.  Although rxm has code to detect simultaneous
connections, the order of CM events as seen by one node may differ
from those seen by another node.  Long story short, one rank may
detect the simultaneous connection request, accepting one, while
rejecting or discarding the other.  But the peer may see that it's
connection request was accepted, transition into a connected state,
then get the connection request from the peer (which the peer has
already decided to discard).  Because transfers may already be
active on the newly formed connection, tearing it down would result
in communication failures.  This has been seen to result in MPI
hanging, since messages have now been lost.

To distinguish between these 2 cases, which require separate
handling, we add a process ID to the CM protocol.  In order to
maintain backwards compatilibity, the PID is added into the
server and client conn_id.  (This means that rxm can now only support
4 billion peers, instead of whatever 2^64 is.)  Existing rxm
implementation will simply reflect the conn_id value back, so we
can mask off the upper bits to convert the conn_id back into an
index.  New implementations of rxm will write their PID into the
upper bits.  If the PID isn't set, the existing behavior will be
used, as we can't distinguish between the above 2 cases.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

shefty commented:
Fixes #7396 (in conjunction with patch already merged)

shefty commented:
bot:aws:retest

j-xiong reviewed (commented):
A few questions related to the backward compatibilty:

> Existing rxm implementation will simply reflect the conn_id value back, so we
can mask off the upper bits to convert the conn_id back into an index.  

Existing implementation doesn't mask the upper bits, does it? Also whether it is upper
bits or lower bits depends on the endianess.

> New implementations of rxm will write their PID into the upper bits.  

again, it depends on the endianess.

> If the PID isn't set, the existing behavior will be used, 

Unset PID is 0.  According to line 672,  the new behavior will be used.

shefty commented:
AFAICT, rxm doesn't perform any byte swapping on the CM message.  So it assumes matching endianness.

It doesn't show up on the patch because the lines didn't change, but the conn_id is cast to an int before being used.  So even though the CM message is u64, anything larger than an int can't be used.  An existing implementation will set conn_id to an index into an array.

I think you're right about the PID check when communicating with an existing peer, and line 672 needs to include a check that the PID is non-zero.  I'll update

shefty commented:
Updated to include pid non-zero check.  Reverted added new pid field -- instead kept u64 conn_id field and added accessors to set/get the pid/index values.
